### ISSUE TITLE
Return 400 Invalid Request for non-ASCII URLs

### ISSUE THREAD

**Issue by sachinrekhi** [django/daphne#150]
Recently launched our product internationally and around the same time started to receive the following UnicodeDecodeError from daphne:

```
Traceback (most recent call last):

File "/daphne/http_protocol.py", line 265, in serverResponse
  "path": self.uri.decode("ascii"),
UnicodeDecodeError: ‘ascii’ codec can't decode byte 0xd0 in position 10: ordinal not in range(128)
```

I'm currently running:

channels==1.1.8
daphne==1.3.0
asgi-redis==1.4.3
asgiref==1.1.2
Django==1.11.8

Any idea what might be causing this?

**Comment by andrewgodwin** [comment]
URIs are meant to be transported as ASCII with any high-byte sequences encoded with percent-encoding, so it's probably a badly behaving HTTP client. You could try changing the decode to `utf8`, but then you might get an error if they're not passing UTF8 but a different character set.

Ideally the best solution here is showing the client a 400 error page. I'll put that on the list for Daphne 2.

**Comment by sachinrekhi** [comment]
OK, sounds good.

I generate links to other parts of my site based on user input that could include high-byte sequences. When I generate a link in the browser on the client, do I need to myself change the url I send to window.location.href to percent-encode high-byte sequences, or is the browser supposed to do that automatically when it redirects the user to the requested url?

**Comment by andrewgodwin** [comment]
I'm not sure if the browser will correctly urlencode it for you without trying it out (it should do, but I bet one or two of them don't)/

**Comment by matthiask** [comment]
I was able to reproduce the problem with `curl -v 'http://127.0.0.1:8000/?äöü'`

I can't make it fail with umlauts before the `?`. Daphne checks early that `path` can be decoded using an ASCII encoding: https://github.com/django/daphne/blob/388bbc5c2412da6d525dfe4edf133fc06d060ab6/daphne/http_protocol.py#L84

However, later when logging the HTTP complete action, the whole `uri` (including the query string) is decoded in https://github.com/django/daphne/blob/388bbc5c2412da6d525dfe4edf133fc06d060ab6/daphne/http_protocol.py#L222 and this leads to an error.

I tried writing a test which makes Daphne fail, but it's surprisingly hard to send requests with non-ASCII URLs. This is what I got: https://github.com/matthiask/daphne/commit/3d81027a511e99fd3e4fabb2687b207398e75c7f

**Comment by andrewgodwin** [comment]
Thanks for that, it gave me enough to reproduce and fix it. I took the inside of your test and put it on the base class instead, meaning the resulting tests look pretty clean!

**Comment by sachinrekhi** [comment]
@andrewgodwin any chance this fix could be backported to the 1.0 series?

**Comment by andrewgodwin** [comment]
Afraid not, the two codebases are substantially different.
